fix: fix exceeded the prepaid gas when upgrade remote contract

fospring opened the pull request.

fospring commented:
When `GAS_FOR_UPGRADE_REMOTE_DEPLOY` is 15T, upgrade remote contract will throw error of E`xceeded the prepaid gas`. Updated `GAS_FOR_UPGRADE_REMOTE_DEPLOY` to 25T would throw same error, So I updated it to 30T.

agileurbanite reviewed (commented):
LGTM, I don't see this negatively impacting other DAOs

agileurbanite commented:
But I believe you'll need @ilblackdragon approval given that he's the main contributor to sputnik atm.

fospring commented:
@ilblackdragon can you pls review this MR?

ilblackdragon commented:
I'm having trouble opening the execution plan for this transaction (https://testnet.nearblocks.io/txns/AR5j8Dz3Ts4ZDKXr1Zd5Eq7VGeS23AmWDtpjhih2wq2m#execution)

Mostly question is why this is needed, because this change may be hiding another issue. Because it the LEFTOVER parameters are to cover costs of calling promise_function_call and returning and it should be not be taking this much gas. It shouldn't even be 15Tg in the first place. 

There is also new API in 4.0 SDK via https://github.com/near/NEPs/pull/264 that allows to not have this at all, at it calculates leftover gas after function already completed: https://github.com/near/near-sdk-rs/issues/740

If we are updating this, let's switch to this API given sputnikdao2 is already on 4.0 sdk.

fospring commented:
> I'm having trouble opening the execution plan for this transaction (https://testnet.nearblocks.io/txns/AR5j8Dz3Ts4ZDKXr1Zd5Eq7VGeS23AmWDtpjhih2wq2m#execution)
> 
> Mostly question is why this is needed, because this change may be hiding another issue. LEFTOVER parameters are to cover costs of calling promise_function_call itself and returning Promise and it should be not be taking this much gas. It shouldn't even be 15Tg in the first place.
> 
> There is also new API in 4.0 SDK via [near/NEPs#264](https://github.com/near/NEPs/pull/264) that allows to not have this at all, at it calculates leftover gas after function already completed: [near/near-sdk-rs#740](https://github.com/near/near-sdk-rs/issues/740)
> 
> If we are updating this, let's switch to this API given sputnikdao2 is already on 4.0 sdk.

fospring commented:
> I'm having trouble opening the execution plan for this transaction (https://testnet.nearblocks.io/txns/AR5j8Dz3Ts4ZDKXr1Zd5Eq7VGeS23AmWDtpjhih2wq2m#execution)
> 
> Mostly question is why this is needed, because this change may be hiding another issue. LEFTOVER parameters are to cover costs of calling promise_function_call itself and returning Promise and it should be not be taking this much gas. It shouldn't even be 15Tg in the first place.
> 
> There is also new API in 4.0 SDK via [near/NEPs#264](https://github.com/near/NEPs/pull/264) that allows to not have this at all, at it calculates leftover gas after function already completed: [near/near-sdk-rs#740](https://github.com/near/near-sdk-rs/issues/740)
> 
> If we are updating this, let's switch to this API given sputnikdao2 is already on 4.0 sdk.

May be there are some trouble for FE in query execution plan:
https://testnet.nearblocks.io/txns/AR5j8Dz3Ts4ZDKXr1Zd5Eq7VGeS23AmWDtpjhih2wq2m#
![image](https://user-images.githubusercontent.com/19633691/216879734-0a3d4890-2372-4762-a9b3-4367c6a533ba.png)

agileurbanite commented:
We need to migrate sputnikdao2 to the latest sdk as it's not even on 4.0 it's currently on 4.0.0-pre.4 which is not even defined as a release version on github:
https://github.com/near/near-sdk-rs/releases?expanded=true&page=2&q=4.0.0-pre.4

But strangely there are rust docs on it:
https://docs.rs/near-sdk/4.0.0-pre.4/near_sdk/index.html

Either way, I'm not sure https://github.com/near/NEPs/pull/264 was even introduced in 4.0.0-pre.5 as it looks like it was pretty of the changes for 4.0.0 stable release here: https://github.com/near/near-sdk-rs/releases/tag/4.0.0

What does this mean?

Well to migrate sputnikdao2 to >= 4.0.0 a few things need to happen:

1. Migrate to the new rust SDK version and update deprecated methods
2. Remove near-sdk-sim and migrate to workspaces-rs

The above is covered in:
https://github.com/near-daos/sputnik-dao-contract/issues/198

Overall, however, my perception of the task is that it seems like a larger effort.

@ilblackdragon do you have concerns with increasing `FACTORY_UPDATE_GAS_LEFTOVER` as a stop gap for now and then during the follow up with updating the SDK we can re-evaluate the need for this as part of that effort.

Also @fospring to @ilblackdragon point, I do believe you need to re-evaluate your implementation as it should not take 15Tg in the first place.  Can you share pseudo code of your implementation or your execution plan in detail?

fospring commented:
> We need to migrate sputnikdao2 to the latest sdk as it's not even on 4.0 it's currently on 4.0.0-pre.4 which is not even defined as a release version on github: https://github.com/near/near-sdk-rs/releases?expanded=true&page=2&q=4.0.0-pre.4
> 
> But strangely there are rust docs on it: https://docs.rs/near-sdk/4.0.0-pre.4/near_sdk/index.html
> 
> Either way, I'm not sure [near/NEPs#264](https://github.com/near/NEPs/pull/264) was even introduced in 4.0.0-pre.5 as it looks like it was pretty of the changes for 4.0.0 stable release here: https://github.com/near/near-sdk-rs/releases/tag/4.0.0
> 
> What does this mean?
> 
> Well to migrate sputnikdao2 to >= 4.0.0 a few things need to happen:
> 
> 1. Migrate to the new rust SDK version and update deprecated methods
> 2. Remove near-sdk-sim and migrate to workspaces-rs
> 
> The above is covered in: #198
> 
> Overall, however, my perception of the task is that it seems like a larger effort.
> 
> @ilblackdragon do you have concerns with increasing `FACTORY_UPDATE_GAS_LEFTOVER` as a stop gap for now and then during the follow up with updating the SDK we can re-evaluate the need for this as part of that effort.
> 
> Also @fospring to @ilblackdragon point, I do believe you need to re-evaluate your implementation as it should not take 15Tg in the first place. Can you share pseudo code of your implementation or your execution plan in detail?

@agileurbanite  @ilblackdragon  My executions flows:
1. create mock accounts:
* [or1007-2.yongchun.testnet](https://explorer.testnet.near.org/transactions/HqxPS1gb1AbPeQ92guTGkWUJFtiPCa5nw94QFvYJG9ME)
* [or1007owner-2.yongchun.testnet](https://explorer.testnet.near.org/transactions/BMLf8E3gC8wczqzPMvxYi2o2GxioqtUgnzdDpk9bFa6d)
2. build dao :
```shell
git clone git@github.com:near-daos/sputnik-dao-contract.git
cargo build --release -p sputnikdao2 --target wasm32-unknown-unknown
```
3. deploy dao contract(`or1007owner-2.yongchun.testnet`, owner of `or1007-2.yongchun.testnet`) by this [transaction](https://explorer.testnet.near.org/transactions/6FwSCPHCxN42RdFpJKnjmsGjWxzieakCX7pifF6XPCMr)
4. deploy smart contract to `or1007-2.yongchun.testnet` with owner `or1007owner-2.yongchun.testnet` by this [transaction](https://explorer.testnet.near.org/transactions/FaZb9E7iPyVqAnnnHcNPVxbbaQeVcXEcByFo7fAqNDN8)
5. store-blob of smart contract, ready to upgrade smart contract of `or1007-2.yongchun.testnet` by this [transaction](https://explorer.testnet.near.org/transactions/BBU8trmh4oYAfVefYcdEog67QXpNBB2XtHNb1mpkjPmi), the store-blob's base58_crypto hash is `Hy8JJwzpRvhnSnEoQnF1WvDU5kZFksAbiScXsDeuzLZg`
6. add proposal of remote upgrade `or1007-2.yongchun.testnet` smart contract, and act proposal:
* [add proposal transaction](https://explorer.testnet.near.org/transactions/EAD8qHmWTfGaCti39DvYdGC1Qu81WUJLb6dDsqAej34k), the proposal id is `0`
* [act proposal transaction](https://explorer.testnet.near.org/transactions/F9CEJmkVTrHgkMJywCB7DsqWfoYULouFVporCKJVHh68) by one approver
* [act proposal transaction](https://explorer.testnet.near.org/transactions/Ej3VfmY3XfbKse6CGi2xhhi7sB6vpyj8gSrUMhw4Gnpr) by another approver 
* with attached 300T gas but failed with error: `ExecutionError": "Exceeded the prepaid gas.`
[transaction execution plan in testnet.nearblocks](https://testnet.nearblocks.io/txns/Ej3VfmY3XfbKse6CGi2xhhi7sB6vpyj8gSrUMhw4Gnpr#execution)
![image](https://user-images.githubusercontent.com/19633691/216921312-65d9936c-cb54-43e4-a0d7-4840e507339f.png)

fospring commented:
> We need to migrate sputnikdao2 to the latest sdk as it's not even on 4.0 it's currently on 4.0.0-pre.4 which is not even defined as a release version on github: https://github.com/near/near-sdk-rs/releases?expanded=true&page=2&q=4.0.0-pre.4
> 
> But strangely there are rust docs on it: https://docs.rs/near-sdk/4.0.0-pre.4/near_sdk/index.html
> 
> Either way, I'm not sure [near/NEPs#264](https://github.com/near/NEPs/pull/264) was even introduced in 4.0.0-pre.5 as it looks like it was pretty of the changes for 4.0.0 stable release here: https://github.com/near/near-sdk-rs/releases/tag/4.0.0
> 
> What does this mean?
> 
> Well to migrate sputnikdao2 to >= 4.0.0 a few things need to happen:
> 
> 1. Migrate to the new rust SDK version and update deprecated methods
> 2. Remove near-sdk-sim and migrate to workspaces-rs
> 
> The above is covered in: #198
> 
> Overall, however, my perception of the task is that it seems like a larger effort.
> 
> @ilblackdragon do you have concerns with increasing `FACTORY_UPDATE_GAS_LEFTOVER` as a stop gap for now and then during the follow up with updating the SDK we can re-evaluate the need for this as part of that effort.
> 
> Also @fospring to @ilblackdragon point, I do believe you need to re-evaluate your implementation as it should not take 15Tg in the first place. Can you share pseudo code of your implementation or your execution plan in detail?

@agileurbanite @ilblackdragon, Thank you for your reminder. I have added new tests that when `FACTORY_UPDATE_GAS_LEFTOVER` is 15Tg, it's enough, so I reverted updation of `FACTORY_UPDATE_GAS_LEFTOVER` by this commit [e2ab7ad036c0a842fd69ba05f7bceb12cc740f84](https://github.com/near-daos/sputnik-dao-contract/pull/197/commits/e2ab7ad036c0a842fd69ba05f7bceb12cc740f84)

fospring commented:
> We need to migrate sputnikdao2 to the latest sdk as it's not even on 4.0 it's currently on 4.0.0-pre.4 which is not even defined as a release version on github: https://github.com/near/near-sdk-rs/releases?expanded=true&page=2&q=4.0.0-pre.4
> 
> But strangely there are rust docs on it: https://docs.rs/near-sdk/4.0.0-pre.4/near_sdk/index.html
> 
> Either way, I'm not sure [near/NEPs#264](https://github.com/near/NEPs/pull/264) was even introduced in 4.0.0-pre.5 as it looks like it was pretty of the changes for 4.0.0 stable release here: https://github.com/near/near-sdk-rs/releases/tag/4.0.0
> 
> What does this mean?
> 
> Well to migrate sputnikdao2 to >= 4.0.0 a few things need to happen:
> 
> 1. Migrate to the new rust SDK version and update deprecated methods
> 2. Remove near-sdk-sim and migrate to workspaces-rs
> 
> The above is covered in: #198
> 
> Overall, however, my perception of the task is that it seems like a larger effort.
> 
> @ilblackdragon do you have concerns with increasing `FACTORY_UPDATE_GAS_LEFTOVER` as a stop gap for now and then during the follow up with updating the SDK we can re-evaluate the need for this as part of that effort.
> 
> Also @fospring to @ilblackdragon point, I do believe you need to re-evaluate your implementation as it should not take 15Tg in the first place. Can you share pseudo code of your implementation or your execution plan in detail?

@agileurbanite @ilblackdragon can we hotfix on this issue?: https://explorer.testnet.near.org/transactions/3xRbdcbdQnC1jSo2arzzQ6bJDNs7PPdrLARmvYFbRL68

frol commented:
It was superseded and fixed in #223
